File: DSA/linked_list.py
# Big O
# append to end is O(N).
# finding an element is O(N)

# linked list
class LinkedListNode:
    def __init__(self, val, next=None):
      self.val = val
      self.next = next

    def __str__(self):
      return f"[{self.val}] -> {self.next}"

# Create nodes: 1->2->3->4->None
node4 = LinkedListNode(4)
node3 = LinkedListNode(3, node4)
node2 = LinkedListNode(2, node3)
head = LinkedListNode(1, node2)

# Alternative one-liner creation
# head = LinkedListNode(1, LinkedListNode(2, LinkedListNode(3, LinkedListNode(4))))

# Print list
print(head)
# [1] -> [2] -> [3] -> [4] -> None
#  ^
# head

# Traverse list
current = head
while current:
    print(current.val, end=" ")  # 1 2 3 4
    current = current.next

# Find node with value x
def find_node(head, x):
    current = head
    while current:
        if current.val == x:
            return current
        current = current.next
    return None

def append_to_end(head, x):
    new_node = LinkedListNode(x)
    current = head
    while current.next:
        current = current.next
    current.next = new_node
    return head

File: DSA/linked_list.rb
class Node
  attr_accessor :value, :next

  def initialize(value)
    @value = value
    @next = nil
  end
end

class LinkedList
  def initialize
    @head = nil
  end

  # Insert at the end
  def append(value)
    new_node = Node.new(value)
    if @head.nil?
      @head = new_node
    else
      current = @head
      current = current.next while current.next
      current.next = new_node
    end
  end

  # Insert at the beginning
  def prepend(value)
    new_node = Node.new(value)
    new_node.next = @head
    @head = new_node
  end

  # Delete a node by value
  def delete(value)
    return if @head.nil?

    if @head.value == value
      @head = @head.next
      return
    end

    current = @head
    while current.next && current.next.value != value
      current = current.next
    end

    current.next = current.next.next if current.next
  end

  # Search for a value
  def search(value)
    current = @head
    while current
      return true if current.value == value
      current = current.next
    end
    false
  end

  # Print the linked list
  def display
    current = @head
    while current
      print "#{current.value} -> "
      current = current.next
    end
    puts "nil"
  end
end

# Example usage
list = LinkedList.new
list.append(10)
list.append(20)
list.append(30)
list.prepend(5)
list.display # Output: 5 -> 10 -> 20 -> 30 -> nil
list.delete(20)
list.display # Output: 5 -> 10 -> 30 -> nil
puts list.search(10) # Output: true
puts list.search(50) # Output: false
